login_server/src/cpp/SingletonManager/EmailManager.cpp

```diff
@@ -141,6 +141,33 @@ int EmailManager::ThreadFunction()
 				mailClientSession.sendMessage(mailMessage);
 					email_sended = true;
 				}
+				catch (Poco::Net::SSLConnectionUnexpectedlyClosedException& ex) {
+					// it is a mad idea to send an email if the email sending failed
+					// errors.sendErrorsAsEmail();
+					// better wait instead and try again, it seems that the strato mail server sometimes discard or connection
+					// MAGIC NUMBER: sleep time if email sending failed (ssl connection was discarded from mailserver)
+					// wait 5 minute for the next try
+					Poco::Thread::sleep(5 * 60 * 1000);
+					// reconnect to mailserver
+					mailClientSession.login();
+					try {
+						mailClientSession.startTLS(ServerConfig::g_SSL_CLient_Context);
+						mailClientSession.login(Poco::Net::SMTPClientSession::AUTH_LOGIN, mEmailAccount.username, mEmailAccount.password);
+					}
```

Comment:
You create duplicate code here - this code part is already present in the top of the function. In general you should avoid that to make it easy to maintain. Now you have to maintain both if something changes in this block, which leads to potential errors in the future.

Furthermore I find it quite weird - or not self explanatory that you call
`mailClientSession.login();` and then `mailClientSession.login(Poco::Net::SMTPClientSession::AUTH_LOGIN, mEmailAccount.username, mEmailAccount.password);`. I did not check this, but I assume that calling `login()` without parameters opens the connections?! It is definitely not self-explanatory which all code should be at all time 🙄

Reply:
"You create duplicate code here - this code part is already present in the top of the function. In general you should avoid that to make it easy to maintain. Now you have to maintain both if something changes in this block, which leads to potential errors in the future." - weiß ich, ich hatte schon überlegt ob ich es in eine separate Funktion auslagere und damit alles noch ein bisschen komplizierter machen soll. Aber ich habe mich dagegen entschieden. Im Regelfall bin ich mir bewusst wenn ich doppelten Code habe und refactore das spätestens wenn ich eine Änderung in diesem Code machen muss.

Reply:
"Furthermore I find it quite weird - or not self explanatory that you call
mailClientSession.login(); and then mailClientSession.login(Poco::Net::SMTPClientSession::AUTH_LOGIN, mEmailAccount.username, mEmailAccount.password);. I did not check this, but I assume that calling login() without parameters opens the connections?! It is definitely not self-explanatory which all code should be at all time 🙄"

Das habe ich einfach aus dem Beispiel übernommen ohne das weiter zu hinterfragen:
https://github.com/pocoproject/poco/blob/master/NetSSL_OpenSSL/samples/Mail/src/Mail.cpp

Reply:
From poco docu for Poco::Net::SecureSMTPClientSession:
```c++
void login(); 
                /// Calls login(hostname) with the current host name.
bool startTLS(Context::Ptr pContext);
		/// Sends a STARTTLS command and, if successful,
		/// creates a secure SSL/TLS connection over the
		/// existing socket connection.
		///
		/// Uses the given Context object for creating
		/// the SSL/TLS connection.
		///
		/// Must be called after login() or login(hostname).
		/// If successful, login() can be called again
		/// to authenticate the user.
		///
		/// Returns true if the STARTTLS command was successful,
		/// false otherwise.
void login(LoginMethod loginMethod, const std::string& username, const std::string& password);
		/// Logs in to the SMTP server using the given authentication method and the given
		/// credentials.
```